code/modules/client/preference_setup/controls/01_keybindings.dm

```diff
@@ -0,0 +1,254 @@
+/datum/preferences
+	/// Custom Keybindings
+	var/list/key_bindings = list()
+	var/hotkeys = TRUE
+
+/// checks through keybindings for outdated unbound keys and updates them
+/datum/preferences/proc/check_keybindings()
+	if(!client)
+		return
+
+	// When loading from savefile key_binding can be null
+	// This happens when player had savefile created before new kb system, but hotkeys was not saved
+	if(!length(key_bindings))
+		key_bindings = deepCopyList(global.hotkey_keybinding_list_by_key) // give them default keybinds too
+
+	var/list/user_binds = list()
+	for (var/key in key_bindings)
+		for(var/kb_name in key_bindings[key])
+			user_binds[kb_name] += list(key)
+	var/list/notadded = list()
+	for (var/name in global.keybindings_by_name)
+		var/datum/keybinding/kb = global.keybindings_by_name[name]
+		if(length(user_binds[kb.name]))
+			continue // key is unbound and or bound to something
+		var/addedbind = FALSE
+		if(hotkeys)
+			for(var/hotkeytobind in kb.hotkey_keys)
+				if(!length(key_bindings[hotkeytobind]) || hotkeytobind == "Unbound") //Only bind to the key if nothing else is bound expect for Unbound
+					LAZYADD(key_bindings[hotkeytobind], kb.name)
+					addedbind = TRUE
+		else
+			for(var/classickeytobind in kb.classic_keys)
+				if(!length(key_bindings[classickeytobind]) || classickeytobind == "Unbound") //Only bind to the key if nothing else is bound expect for Unbound
+					LAZYADD(key_bindings[classickeytobind], kb.name)
+					addedbind = TRUE
+		if(!addedbind)
+			notadded += kb
+
+	if(length(notadded))
+		addtimer(CALLBACK(src, .proc/announce_conflict, notadded), 5 SECONDS)
+
+/datum/preferences/proc/announce_conflict(list/notadded)
+	to_chat(client, "<span class='userdanger'>KEYBINDING CONFLICT!!!\n\
```

Comment:
`!` is sufficient, multiple exclamation marks just looks silly.
Do we have a `userdanger` class?

Reply:
Probably no. I will replace it with better span.